src/platform/Linux/OTAImageProcessorImpl.cpp

```diff
@@ -134,7 +135,8 @@ void OTAImageProcessorImpl::HandleApply(intptr_t context)
     OTARequestorInterface * requestor = chip::GetRequestorInstance();
     if (requestor != nullptr)
     {
-        requestor->NotifyUpdateApplied(imageProcessor->mHeader.softwareVersion);
+        // TODO: Use software version from Configuration Manager
+        requestor->NotifyUpdateApplied(imageProcessor->mSoftwareVersion);
```

Comment:
The NotifyUpdateApplied must be called from the image version running, not at the time of receiving "apply". Otherwise you basically are telling a provider that you actually succeeded in running the new version, when you only succeeded doing some internal switching to the new image.

Overall the call chain:

- OnApplyUpdateResponse()
  - UpdateConfirmed()
    - imageProcessor->Apply()
      - requestor->NotifyUpdateApplied()
      
is an invalid sequence.

Looking at the spec:

> [[OTA-NotifyUpdateApplied-SoftwareVersion]]
> ==== SoftwareVersion field
> The SoftwareVersion included in the request payload SHALL provide the same value as the <<ref_BasicInfoCluster_SoftwareVersion,`SoftwareVersion`>> attribute in the invoking OTA Requestor's <<ref_BasicInformationCluster,Basic Information Cluster>>, and SHOULD be consistent with the value representing a new version running on the Node invoking the command.
> 
> ==== When Generated
> The `NotifyUpdateApplied` command SHOULD be invoked in the following two circumstances:
> 
> . An OTA Requestor has just successfully applied a Software Image it had obtained from a previous <<OTA-Command-QueryImageResponse,`QueryImage` response>>.
> . An OTA Requestor has just successfully applied a Software Image it had obtained through means different than those of this Cluster.
> 
> An OTA Provider MAY use the state of invocation of this command to help track the progress of update for OTA Requestors it knows require a new OTA Software Image. However, due to the possibility that an OTA Requestor MAY never come back (e.g. device removed from Fabric altogether, or a critical malfunction), an OTA Provider SHALL NOT expect every OTA Requestor to invoke this command for correct operation of the OTA Provider.
> 
> This command SHALL be considered optional and SHALL not result in reduced availability of the OTA Provider functionality if OTA Requestors never invoke this command.
> 

I can see how it can be understood to be the against the current running version, but the intent was truly to allow OTA Providers to keep track of whether there was successful application of OTA that got to a running state.

Reply:
I know, but I didn't want to modify the current behavior in this unrelated PR. The Linux implementation currently, doesn't do the application restart, so it uses this hacky solution.

Reply:
This particular choice was made because it would not be desirable or practical to actually restart the Reference OTA Requestor app on Linux. On the other hand, given that the app is used as a test tool it was desirable for it to be able to issue `NotifyUpdateApplied` . We should have added a comment discussing all of this.